### ISSUE TITLE
sonar_runner console output  analysing Unit Tests 

### ISSUE THREAD

**Issue by Benoitlz** [SonarOpenCommunity/sonar-cxx#520]
Unit tests analysis seems to be handled differently from the main code.
When the main code is analyzed, sonar_runner console output produces only once the block of WARN and INFO at the beginning.
For the Unit tests, it produces this block (which in my case is 230 lines) for every file that has a parsing error. Please produce it only once at the beginning of the unit tests, this information does not change.
(in example below starts at '14:15:21.390 WARN')
14:15:21.091 INFO  - Sensor CxxXunitSensor
14:15:21.109 INFO  - Processing report 'M:\wcs_simulation_QA_373\debug\GtestReports.xml'
14:15:21.387 INFO  - Parsing 'xUnit' format
14:15:21.388 INFO  - Processing in 'detailled mode' i.e. with provideDetails=true
14:15:21.390 WARN  - the include root 'C:\Program Files (x86)\Intel\Composer XE 2015\compiler\include\intel64' doesn't exist
14:15:21.391 WARN  - the include root 'C:\Program Files\Java\jre1.8.0_45\include\win32' doesn't exist
14:15:21.392 WARN  - the include root 'C:\Program Files\Java\jre1.8.0_45\INCLUDE' doesn't exist
14:15:21.395 INFO  - storing external macro: 'USE_WIX39 -> '''
14:15:21.395 INFO  - storing external macro: 'ON_WM_KILLFOCUS() -> '''
14:15:21.395 INFO  - storing external macro: '_RWTOOLSDLL -> '''

Also code that seems to parse OK for the main code does not parse in the unit tests, e.g.:
namespace MINE { namespace WCS { namespace Simulator { namespace Displacement {

**Comment by jmecosta** [comment]
Thats expected, test code is not imported into the statistics. In current case they are parsed to try to map test cases to source.

**Comment by Benoitlz** [comment]
my point here was:
1) can the console output from sonar_runner be reduced ?
2) can the code that parses OK in main project parse as well when in unit tests (the same syntax fails in the second case) ?
Thanks and Regards.

**Comment by Benoitlz** [comment]
I'm not a java developer, but looking at CxxPreprocessor.java -> for (String define : conf.getDefines()) , do we really expect to have the definitions from one unit test to the next to change ?
If not the case, this could be done once and reused, improving parsing performance (and minimizing Log,info output).

**Comment by guwirth** [comment]
> I'm not a java developer, but looking at CxxPreprocessor.java -> for (String define : conf.getDefines()) , do we really expect to have the definitions from one unit test to the next to change ?

@Benoitlz It is in the ctor, means it is called only once per CxxPreprocessor instance. What is the problem with this?

**Comment by Benoitlz** [comment]
Would this means that every time there is a a parsing issue with an Unit Test, a new coprocessor instance is created ? 
This does not occur with the main code.
In my case, there is more than 200 lines  'storing external macro'. Now this get repeated for each UnitTest file (that has a parsing error?). This just hides possibly useful information and probably slows down the parsing.

**Comment by jmecosta** [comment]
This can be improved but its not critical. If you want to can use the other
mode that skip the parsing given that you augment the information in the
reports to include the location of each test. See the wiki for details and
the integration test for some examples

**Comment by guwirth** [comment]
Problem is https://github.com/wenns/sonar-cxx/blob/master/sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/xunit/CxxXunitSensor.java where in buildLookupTables the method CxxAstScanner.scanSingleFileConfig is called for each file. A solution could be to split CxxAstScanner.scanSingleFileConfig into calling 'public static AstScanner<Grammar> create(CxxConfiguration conf, SquidAstVisitor<Grammar>... visitors)' outside the loop and create a new 
CxxAstScanner.scanSingleFileScanner handing over the file and the existing scanner (https://github.com/wenns/sonar-cxx/blob/59399627529ffc09a56f89345e910bd40bc65aa0/cxx-squid/src/main/java/org/sonar/cxx/CxxAstScanner.java).

**Comment by guwirth** [comment]
Performance quick win: I will provide a solution

**Comment by Benoitlz** [comment]
Tested Unit Tests Parsing with 0.9.4-RC1.
The previous issue with 0.9.3 macro parsing in log file is now solved.
There is a new issue in log file now: each syntax error generates 20 to 40 times the same log warning, e.g.:
19:45:14.637 WARN  - [M:\wcs_simulation_QA_385\...\FluidDensityInterfaceTest.cpp:115]: syntax error, skip 'testing'

As I have some issues parsing Google Tests, sonar_runner generates a 1.9 GB log file!

**Comment by guwirth** [comment]
@Benoitlz This is a new problem please create a new issue. 

What is your expectation? Adding the message only once per line number, per cpp file, ... ?

**Comment by Benoitlz** [comment]
Open issue:
 Testing Unit Tests Parsing with 0.9.4-RC1 - repeated log messages #535

**Comment by Benoitlz** [comment]
open issue: Testing Unit Tests Parsing with 0.9.4-RC1 - repeated log messages #535

**Comment by guwirth** [comment]
seems to work for others
